core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/function/NumberSelection.java

```diff
@@ -19,172 +19,57 @@
 import io.cdap.cdap.api.data.format.StructuredRecord;
 import io.cdap.cdap.api.data.schema.Schema;
 
-import java.util.List;
-import javax.annotation.Nullable;
-
 /**
  * Base class for number based selection functions.
  * Allows subclasses to implement typed methods instead of implementing their own casting logic.
  * Guarantees that only methods for one type will be called for each aggregate. For example,
- * if {@link #operateOn(StructuredRecord)} is called, only {@link #operateOnInt(int, StructuredRecord)} will be called.
+ * if {@link #select(StructuredRecord, StructuredRecord)} is called,
+ * only {@link #combineInt(int, int)} will be called.
  */
 public abstract class NumberSelection implements SelectionFunction {
-  private final SelectionFunction delegate;
   private final String fieldName;
+  private final Schema.Type fieldType;
 
-  public NumberSelection(final String fieldName, @Nullable Schema fieldSchema) {
+  public NumberSelection(String fieldName, Schema fieldSchema) {
     this.fieldName = fieldName;
-    // if schema is not known before we start getting records, just use doubles.
-    if (fieldSchema == null) {
-      delegate = new SelectionFunction() {
-        @Override
-        public List<StructuredRecord> getSelectedRecords() {
-          return getRecords();
-        }
+    this.fieldType = fieldSchema.isNullable() ? fieldSchema.getNonNullable().getType() : fieldSchema.getType();
+  }
 
-        @Override
-        public void beginFunction() {
-          startDouble();
-        }
+  protected abstract int combineInt(int val1, int val2);
 
-        @Override
-        public void operateOn(StructuredRecord record) {
-          Number number = record.get(fieldName);
-          if (number != null) {
-            operateOnDouble(number.doubleValue(), record);
-          }
-        }
-      };
-      return;
-    }
+  protected abstract long combineLong(long val1, long val2);
 
-    final boolean isNullable = fieldSchema.isNullable();
-    Schema.Type fieldType = isNullable ? fieldSchema.getNonNullable().getType() : fieldSchema.getType();
-    switch (fieldType) {
-      case INT:
-        delegate = new SelectionFunction() {
-          @Override
-          public List<StructuredRecord> getSelectedRecords() {
-            return getRecords();
-          }
+  protected abstract float combineFloat(float val1, float val2);
```

Comment:
these methods should be called 'compare' instead of 'combine', and they should all return an int.

Reply:
good catch, the implementation is based on compare, I forgot to change this.